qupulse/hardware/awgs/decadac.py

```diff
@@ -0,0 +1,440 @@
+from numbers import Real
+from typing import Tuple, Callable, Optional, Sequence, Union, Dict, Mapping, Set
+import serial
+
+from qupulse import ChannelID
+from qupulse._program._loop import Loop, to_waveform
+from qupulse.hardware.awgs.base import AWG
+
+'''upload general plan:
+    get timeline, with each instance being an index in an array
+    convert every voltage value to a decadac value
+    look for ramps '''
+def convert_voltage(v) -> int:
+    return int((v + 10) / 20 * 65535)
+
+#TODO delete this at one point
+'''class Single:
+
+    def __int__(self, voltage, duration):
+        self.d = voltage
+        self. duration = duration
+
+    def get_ascii(self):
+        return 'D' + str(self.d)
+
+class Ramp:
+
+    def __init__(self, start_v: int, end_v: int, timestep: int, steps: int):
+        self.start = start_v
+        self.end = end_v
+        self.timestep = timestep
+        self.steps = steps
+        self.lim = 'U'
+
+    def get_ascii(self):
+        if self.start > self.end:
+            self.lim = 'L'
+        s = int((self.end - self.start) / self.steps * 65536)
+        return self.lim + str(self.end) + ';T' + str(self.timestep) + ';S' + str(s) + ';'
+
+    def reset(self):
+        edge = 65535
+        if self.start > self.end:
+            edge = 0
+        return self.lim + str(edge)
+
+    def increase_step(self):
+        self.steps += 1
+
+    @property
+    def endv(self):
+        return self.end
+
+    @endv.setter
+    def endv(self, newv):
+        self.end = newv
+
+    @property
+    def T(self):
+        return self.timestep
+
+    @property
+    def duration(self):
+        return self.steps * self.timestep
+
+
+
+
+class Iteration:
+
+    def __int__(self, location: int, iteration_number: int, code_block: int):
+        self._loc = location
+        self._num = iteration_number
+        self._block = code_block
+
+    @property
+    def loc(self):
+        return self._loc
+
+    @property
+    def num(self):
+        return self._num
+
+    @property
+    def block(self):
+        return self._block
+
+    # TODO see if you need to also return ;
+    def point(self) -> str:
+        return 'A' + str(self.loc)
+
+    def set_itr(self) -> str:
+        return 'P' + str(self.num)
+
+    @staticmethod
+    def decrement():
+        return '+-1'
+        '''
+
+'''class Initial:
+    def __int__(self, voltage, duration):
+        self.voltage = voltage
+        self.duration = duration'''
+class Stay:
+    def __init__(self, voltage):
+        self.voltage = voltage
+class Ramp:
+    def __init__(self, v_step):
+        self.v_step = v_step
+        
+class Assign:
+
+    def __init__(self, voltage, channel: tuple):
+        self._voltage = voltage
+        self._channel = channel # in the form of (B,C) where B is the board and C is the channel. Possible values: B 0-4, C 0-3
+
+
+    def get_ascii(self):
+        command = 'B' + str(self._channel[0]) + ';C' + str(self._channel[1]) + ';'

```

Comment:
Use the python format string literals where possible. It makes reading the ASCII commands in the code easier and is generally faster.
  ```python
command = f"D{self._channel[0]};C{self._channel[1]};"
# vs
command  = 'B' + str(self._channel[0]) + ';C' + str(self._channel[1]) + ';'
# although the best would be
command = f"D{self.block};C{self.channel};"
```